### PR TITLE
Add solutions for the 4th assignment

### PR DESCRIPTION
**Pull request by nfadeeva**

### CODE REVIEW

File: nfadeeva/assignment4/tests.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                       [True, True, False, False],
+                       [False, False, True, False],
+                       [False, False, True, False]]
+        num_rows = 4
+        num_cols = 4
+        self.assertEqual(count_islands(num_rows, num_cols, input_array), 3)
+
+    def test_empty_array(self):
+        input_array = []
+        num_rows = 0
```
Comment (ariava) [inline nfadeeva/assignment4/tests.py]: nit: you can also write:
```
num_rows, num_cols = 0, 0
```
This doesn't change anything, it's just more compact.

File: nfadeeva/assignment4/count_islands.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from typing import List
+
+
+def neighbors(input_array, num_rows, num_cols, x, y):
```
Comment (ariava) [inline nfadeeva/assignment4/count_islands.py]: Maybe for clarity rename input_array to map?

File: nfadeeva/assignment4/count_islands.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+def dfs(num_rows, num_cols, input_array, x, y):
+
+    # mark the current tile as visited
+    input_array[x][y] = False
+    # start dfs from every neighbour of the current tile
+    for next_x, next_y in neighbors(input_array, num_rows, num_cols, x, y):
+        dfs(num_rows, num_cols, input_array, next_x, next_y)
+
+
+def count_islands(num_rows, num_cols, input_array: List[bool]):
```
Comment (ariava) [inline nfadeeva/assignment4/count_islands.py]: Is input_array a list of bool or a list of lists of bool?
Try running your code through mypy:
```
$ mypy count_islands.py
```

**Review by ariava — APPROVED** [review]
Good job! Please take a look at the comments I left.

**Review by ariava — CHANGES_REQUESTED** [review]
Clicked "approve" by mistake :/ - have a look at my previous comments.

File: nfadeeva/assignment4/count_islands.py (hunk 2 above)
Comment (nfadeeva) [reply nfadeeva/assignment4/count_islands.py]: @ariava I fixed all places and renamed input_array to tiles because map is the built-in name in python)

**Review by ariava — APPROVED** [review]
Super cool!

**Pull request merged**